bokehjs/src/lib/models/text/utils.ts

```diff
@@ -1,7 +1,118 @@
+import {PlainText} from "./plain_text"
+import {TeX} from "./math_text"
 import {isString} from "core/util/types"
-import {TeX} from "models/text/index"
 import {BaseText} from "./base_text"
 
+type Delimiter = {
+  start: string
+  end: string
+  inline: boolean
+  nextIndex?: number
+}
+
+export function find_math_parts(mathstring: string): (PlainText | TeX)[] {
```

Comment:
Same comment as above, this is incredibly complicated and opaque manual text processing code. I don't recommend regex often but I don't see how this isn't a case to use non-greedy expressions for find all the pairs, more efficiently and with less code to maintain.

Reply:
So this is a shallow copy of what MathJax does. It is not a simple function to look for pair delimiters, it interprets them in order and save their positions on the text input. Maybe should have a better name.

Reply:
can you link to the mathjax code? If this is truly what they do then I will withdraw my objection, but then we should leave comments with that link to the reference code we are drawing from.

Reply:
https://github.com/mathjax/MathJax-src/blob/master/ts/input/tex/FindTeX.ts#L197

Reply:
I'm struggling a bit to see the direct connection. The MathJax code does use regexes (several calls to [`RegExp.exec`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec) throughout) but does not use `indexOf`, `map`, `filter`, or `sort` at all for searching. I guess we can revisit later as long as there is test coverage.

 But speaking of that, this function would be better if was decoupled from Bokeh models. It should *find* math parts, not find parts and also *convert* math parts. This function should return something like MathJax's `ProtoItem`, i.e. a thing that says *"Here's the text I found, here is what kind of text it is, here are the indices into the original text where I found it"* Then callers of `find_math_parts` can use that to convert as appropriate. That will be more straightforward to test (since checks can be by value, not by types).